sotodlib/hwp/hwp.py

```diff
@@ -6,429 +6,577 @@
+def extract_hwpss(aman, signal=None, hwp_angle=None,
+                  bin_signal=True, bins=3600,
+                  lin_reg=True, modes=[1, 2, 3, 4, 6, 8],
+                  apply_prefilt=True, prefilt_cutoff=1.0,
+                  mask_flags=True,
+                  add_stats_to_aman=True, hwpss_stats_name='hwpss_stats',
+                  add_extracted_to_aman=True, hwpss_extract_name='hwpss_extract'):
+    """
+    Extracts HWP synchronous signal (HWPSS) from a time-ordered data (TOD) using linear regression or curve-fitting.
+
+    Parameters
+    ----------
+    aman : AxisManager object
+        The TOD to extract HWPSS from.
+    signal : array-like, optional
+        The TOD signal to use. If not provided, `aman.signal` will be used.
+    hwp_angle : array-like, optional
+        The HWP angle for each sample in `aman`. If not provided, `aman.hwp_angle` will be used.
+    bin_signal : bool, optional
+        Whether to bin the TOD signal into HWP angle bins before extracting HWPSS. Default is `True`.
+    bins : int, optional
+        The number of HWP angle bins to use if `bin_signal` is `True`. Default is 3600.
+    lin_reg : bool, optional
+        Whether to use linear regression to extract HWPSS from the binned signal. If `False`, curve-fitting will be used instead.
+        Default is `True`.
+    modes : list of int, optional
+        The HWPSS harmonic modes to extract. Default is [1, 2, 3, 4, 6, 8].
+    apply_prefilt : bool, optional
+        Whether to apply a high-pass filter to signal before extracting HWPSS. Default is `True`.
+    prefilt_cutoff : float, optional
+        The cutoff frequency of the high-pass filter, in Hz. Only used if `apply_prefilt` is `True`. Default is 1.0.
+    mask_flags : bool, optional
+        Whether to mask out flagged samples before extracting HWPSS. Default is `True`.
+    add_stats_to_aman : bool, optional
+        Whether to add the extracted HWPSS statistics to `aman` as new axes. Default is `True`.
+    hwpss_stats_name : str, optional
+        The name to use for the new axis containing the HWPSS statistics if `add_stats_to_aman` is `True`. Default is 'hwpss_stats'.
+    add_extracted_to_aman : bool, optional
+        Whether to add the extracted HWPSS to `aman` as a new signal axis. Default is `True`.
+    hwpss_extract_name : str, optional
+        The name to use for the new signal axis containing the extracted HWPSS if `add_extracted_to_aman` is `True`. Default is 'hwpss_extract'.
 
+    Returns
+    -------
+    hwpss_stats : AxisManager object
+        The extracted HWPSS and its statistics.
+    """
 
-def func_hwpss_2f(x, A2, phi2):
-    return A2 * np.cos(2 * x + phi2)
+    if signal is None:
+        if apply_prefilt:
+            filt = tod_ops.filters.high_pass_sine2(cutoff=prefilt_cutoff)
+            signal = np.array(tod_ops.fourier_filter(
+                aman, filt, detrend='linear', signal_name='signal'))
```

Comment:
We've recently been timing things and finding that detrending within the fourier filtering is MUCH slower than detrending in place (6s vs 120s for 1.2 hours of data). If `signal=None` and `apply_prefilt=True` are the default way of calling this function then we might want to re-evaluate the order of operations here.

Reply:
Thanks Katie. I did not know that.
Then, I set an argument of `prefilt_detrend` which specifies whether apply detrend in prefiltering.
I leave the default that `prefilt_detrend='linear'` , but you can specify `prefilt_detrend=None` if the data is already detrended.

Reply:
@kmharrington I'm going to say this is resolved with the fix from @tterasaki if you want to open back up you can.